### DISCUSSION TITLE
waypaper systemd service

### DISCUSSION THREAD

**Discussion by sullyj3** [YaLTeR/niri#468] [Q&A]
I was using the swaybg systemd service recommended on [the wiki](https://github.com/YaLTeR/niri/wiki/Example-systemd-Setup). I decided I wanted to be able to set my wallpaper graphically, so I installed waypaper (basically nitrogen for wayland). `waypaper --restore` from the cli works great.

```
▷ waypaper --restore
swaybg: no process found
Sent command to set wallpaper was set with swaybg
2024-06-25 18:12:09 - [main.c:282] Found config * for output HDMI-A-1 ((null))
```

However, I'm running into issues trying to turn it into a systemd service. I created `waypaper-restore.service` in `~/.config/systemd/user`, and linked it into `niri.service.wants`

```systemd
[Unit]
PartOf=graphical-session.target
After=graphical-session.target
Requisite=graphical-session.target

[Service]
Type=oneshot
ExecStart=/usr/bin/env waypaper --restore
```

It seems like it's maybe running too early or something?
```
▷ journalctl --user -u waypaper-restore.service|tail -5
Jun 25 17:41:14 phrygian systemd[1557]: Starting waypaper-restore.service...
Jun 25 17:41:14 phrygian env[3203]: swaybg: no process found
Jun 25 17:41:14 phrygian env[3207]: 2024-06-25 17:41:14 - [main.c:562] Unable to connect to the compositor. If your compositor is running, check or set the WAYLAND_DISPLAY environment variable.
Jun 25 17:41:14 phrygian env[3156]: Sent command to set wallpaper was set with swaybg
Jun 25 17:41:14 phrygian systemd[1557]: Finished waypaper-restore.service.
```

As a long shot I tried setting `Environment=WAYLAND_DISPLAY=wayland-1` (the output of `$echo WAYLAND_DISPLAY`) under `[Service]`, but it didn't work. That makes sense, niri is setting it anyway, so it should be available when the service runs. 

Any ideas how to get this working?

**Comment by kls1991** [comment]
Try `After=niri.service` ?
I have this line in my waybar service

**Reply by sullyj3** [reply]
hmm, I'll give that a shot

**Comment by YaLTeR** [comment]
Not sure, if swaybg works then the environment should be there.

**Reply by sullyj3** [reply]
yeah, I found this very confusing for that reason

**Comment by jgregoire** [comment]
Mine looks like

```
[Unit]
PartOf=graphical-session.target
After=graphical-session.target
Requisite=graphical-session.target

[Service]
ExecStart=niri msg action spawn -- waypaper --restore --random
```

This has worked reliably for me so far.
